recursion/writing-recursive-code.md

```diff
@@ -0,0 +1,183 @@
+
+### Describe the Repeated Smaller Problem
+
+The problem states that the function "computes the sum of all natural number up to and including `num`."
+
+<!-- Question 1 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: paragraph
+* id: 7lIVkd
+* title: Writing Recursive Code
+##### !question
+
+Consider the `sum_natural_numbers` problem.
+
+What is one way we describe the repeated problem? What is the problem that gets divided into smaller versions of itself?
+
+##### !end-question
+##### !placeholder
+
+The repeated problem for sum_natural_numbers is...
+
+##### !end-placeholder
+##### !explanation
+
+We might describe the repeated problem like so:
+
+The repeated problem is addition. We need to add the sum of all natural numbers calculated so far and a new number.
+
+##### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+### Determine the Information the Smaller Problem Needs
+
+Each smaller version of the problem needs an input number, `num`.
+
+Each recursive call should bring the number closer to the base case, `1`. Because `1` is the smallest natural number, and all other integers are greater than `1`, we should start our recursion from a larger integer and decrement `num`.
+
+By decrementing, we can predict that `num` will eventually become `1` and end the recursion.
+
+### Determine the Information We Want Back
+
+The value we need to `return` from each call is the sum of all natural numbers up to and including `num`.
+
+### Create Example Tables
+
+The following table represents the system stack when calling `sum_natural_numbers(4)`. The top of the stack is the first row, `sum_natural_numbers(1)`. This function call is at the top of the stack because it is the base case, so it is the final function call. With stacks, the last item pushed in is the first item popped out (LIFO!).
+
+| Function Call            | `num` | Return value |
+| ------------------------ | ----- | ------------ |
+| `sum_natural_numbers(1)` | `1`   | `1`          |
+| `sum_natural_numbers(2)` | `2`   | `3`          |
+| `sum_natural_numbers(3)` | `3`   | `6`          |
+| `sum_natural_numbers(4)` | `4`   | `10`         |
+
+### Putting It Together
+
+Here is an example implementation of `sum_natural_numbers`:
+
+```python
+def sum_natural_numbers(num):
+    print("Calling sum_natural_numbers! num:", num)
+
+    # Base Case: When num is 1, the sum is 1
+    if num == 1:
+        return 1
+
+    # Return the sum of all natural numbers so far plus the
+    # current num.
+    # Decrementing num by 1 brings us closer to the base case.
+    return sum_natural_numbers(num-1) + num
+
+
+print("The sum of all natural numbers up to and including 4 is",
+      sum_natural_numbers(4))
+
+print("-------------")
+
+print("The sum of all natural numbers up to and including 5 is",
+      sum_natural_numbers(5))
+
+print("-------------")
+
+print("The sum of all natural numbers up to and including 1 is",
+      sum_natural_numbers(1))
+```
+
```

Comment:
designing recursion is hard, I can imagine many students getting here and still feeling very confused. I want to validate these feelings and provide some suggestions.
```suggestion

### !callout-info

## Writing Recursive Code Takes Practice

After reading through this lesson, it's okay if it's not entirely clear how we arrive at recursive algorithms. Writing them often depends on having a bit of an "Aha!" moment. Sometimes, charting the data in an iterative solution can reveal a self-similarity in the calculations that provides that moment of recognition. Other times, thinking about a problem in terms like, "I want to do _X_, but to do so, first I need to do _Y_," where _Y_ is somehow a "smaller" version of _X_ can also be helpful.

<br />

In any case, we shouldn't get discouraged if the that key recursive step doesn't immediately leap out at us. Practice will help us develop the knack for seeing it!

### !end-callout

```